src/utils/supergraph_config.rs

```diff
@@ -0,0 +1,1169 @@
+/// [`SupergraphConfig`] representation
+#[derive(Clone, Debug)]
+pub struct RemoteSubgraphs(SupergraphConfig);
+
+impl RemoteSubgraphs {
+    /// Fetches [`RemoteSubgraphs`] from Studio
+    pub fn fetch(
+        client: &StudioClient,
+        federation_version: &FederationVersion,
+        graph_ref: &GraphRef,
+    ) -> RoverResult<RemoteSubgraphs> {
+        let subgraphs = subgraph::fetch_all::run(
+            SubgraphFetchAllInput {
+                graph_ref: graph_ref.clone(),
+            },
+            client,
+        )?;
+        let subgraphs = subgraphs
+            .iter()
+            .map(|subgraph| {
+                (
+                    subgraph.name().clone(),
+                    SubgraphConfig {
+                        routing_url: subgraph.url().clone(),
+                        schema: SchemaSource::Sdl {
+                            sdl: subgraph.sdl().clone(),
+                        },
+                    },
+                )
+            })
+            .collect();
+        let supergraph_config = SupergraphConfig::new(subgraphs, Some(federation_version.clone()));
+        let remote_subgraphs = RemoteSubgraphs(supergraph_config);
+        Ok(remote_subgraphs)
+    }
+
+    /// Provides a reference to the inner value of this representation
+    pub fn inner(&self) -> &SupergraphConfig {
+        &self.0
+    }
+}
+
+pub fn get_supergraph_config(
+    graph_ref: &Option<GraphRef>,
+    supergraph_config_path: &Option<FileDescriptorType>,
+    federation_version: &FederationVersion,
+    client_config: StudioClientConfig,
+    profile_opt: &ProfileOpt,
+) -> Result<Option<SupergraphConfig>, RoverError> {
+    // Read in Remote subgraphs
+    let remote_subgraphs = match graph_ref {
+        Some(graph_ref) => {
+            let studio_client = client_config.get_authenticated_client(profile_opt)?;
+            Some(RemoteSubgraphs::fetch(
+                &studio_client,
+                federation_version,
+                graph_ref,
+            )?)
+        }
+        None => None,
+    };
+
+    // Read in Local Supergraph Config
+    let supergraph_config = if let Some(file_descriptor) = &supergraph_config_path {
+        Some(resolve_supergraph_yaml(
+            file_descriptor,
+            client_config,
+            profile_opt,
+        )?)
+    } else {
+        None
+    };
+
+    // Merge Remote and Local Supergraph Configs
+    let supergraph_config = match remote_subgraphs {
+        Some(remote_subgraphs) => match supergraph_config {
+            Some(supergraph_config) => {
+                let mut merged_supergraph_config = remote_subgraphs.inner().clone();
+                merged_supergraph_config.merge_subgraphs(&supergraph_config);
+                Some(merged_supergraph_config)
+            }
+            None => Some(remote_subgraphs.inner().clone()),
+        },
+        None => supergraph_config,
+    };
+    Ok(supergraph_config)
+}
```

Comment:
this was kind of hard for me to follow; not a blocker, but wondering if `.and_then()` or `if let` would make it clearer? probs not worth worrying about given how quickly we'd like to get this in, though

Reply:
Yer I agree it's confusing, went down a slightly different path and did a match on a tuple of the remote and local subgraphs because that's really what this is saying.